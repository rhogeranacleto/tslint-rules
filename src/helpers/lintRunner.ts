import { Configuration, Linter, LintResult } from 'tslint';

interface IMaria {
	src: string;
	// tslint:disable-next-line:no-any
	rule: any;
	fileName?: string;
	fix?: boolean;
}

export function lintRunner(data: IMaria) {

	const { src, rule, fileName = `/tmp/test-${Date.now()}.snap`, fix = false } = data;

	const linter = new Linter({ fix });

	linter.lint(
		fileName,
		src,
		Configuration.parseConfigFile({
			rules: {
				[rule.name || rule]: [true]
			},
			rulesDirectory: 'src'
		})
	);

	return linter.getResult();
}

export function getReplacements(result: LintResult) {

	if (result.fixes) {

		return result.fixes.map(fix => {

			const fixJson = fix.toJson().fix;

			if (fixJson) {

				if (Array.isArray(fixJson)) {

					return fixJson.map(replace => replace.innerText);
				}

				return [fixJson.innerText];
			}
		});
	}
}