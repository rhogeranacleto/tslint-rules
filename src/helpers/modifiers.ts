export function removeLastEmptyLineOfTheBlock(block: string) {

	const lines = block.split('\n');
	const bracket = lines.pop();

	for (let i = lines.length - 1; i > 0; i--) {

		if (lines[i].trim()) {

			break;
		}

		lines.pop();
	}

	lines.push(bracket as string);

	return lines.join('\n');
}