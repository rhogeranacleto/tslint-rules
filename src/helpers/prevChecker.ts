// tslint:disable-next-line:no-implicit-dependencies
import { getPreviousStatement, getPreviousToken } from 'tsutils';
import ts from 'typescript';

export function prevStatementChecker(node: ts.Statement, sourceFile: ts.SourceFile) {

	const prev = getPreviousStatement(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

export function prevTokenChecker(node: ts.Node, sourceFile: ts.SourceFile) {

	const prev = getPreviousToken(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

function checkPrevLine(node: ts.Node, sourceFile: ts.SourceFile, prev: ts.Node) {

	const startLine = ts.getLineAndCharacterOfPosition(sourceFile, node.getStart()).line;
	const prevLine = ts.getLineAndCharacterOfPosition(sourceFile, prev.getEnd()).line;

	return prevLine === (startLine - 1);
}