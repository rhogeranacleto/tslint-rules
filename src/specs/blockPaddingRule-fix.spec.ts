import { getReplacements, lintRunner } from '../helpers/lintRunner';

const RULE = 'block-padding';

describe('blockPadding Rule', () => {

	it.each([
		[`if (true) {

			console.log(true);

			joao();


		}`],
		[`class A {
			public b() {

				console.log(1);

			}
		}`],
		[`it('jj', () => {

			expect(result.errorCount).toBe(1);

		});`],
		[`function maria(){

			console.log('true');

		}`],
		[`if (true) {

			console.log(true);
		} else {

			console.log(false)

		}`],
		[`if (true) {

			console.log(true);

		} else {

			console.log(false)

		}`],
		[`if (true) {

			console.log(true);

		} else if (false) {

			console.log(false)

		}`],
		[`done(() => {

			return 'maria';

		})`],
		[`for (let i = 0; i < 10; i++) {

			console.log(i);

		}`],
		[`for (let test in testes) {

			console.log(test);

		}`],
		[`for (let test of testes) {

			console.log(test);

		}`],
		[`try {

			console.log(0/0);

		} catch(e){

			console.log(0)

		}`],
		[`if (true) {
			console.log(true);
		}`],
		[`class A {
			constructor() {
				console.log(1);
			}
		}`],
		[`class A {
			public b() {
				console.log(1);
			}
		}`],
		[`it('jj', () => {
			expect(result.errorCount).toBe(1);
	});`],
		[`function maria(){
		console.log('true');
	}`],
		[`if (true) {

		console.log(true);
	} else {
		console.log(false)
	}`],
		[`if (true) {
		console.log(true);
	} else {
		console.log(false)
	}`],
		[`if (true) {
		console.log(true);
	} else if (false) {
		console.log(false)
	}`],
		[`for (let i = 0; i < 10; i++) {
		console.log(i);
	}`],
		[`for (let test in testes) {
		console.log(test);
	}`],
		[`for (let test of testes) {
		console.log(test);
	}`],
		[`try {
		console.log(0/0);
	} catch(e){
		console.log(0)
	}`],
		[`done(() => {
		return 'maria';
	})`]
	])('%s should throw error for line after method', src => {

		const result = lintRunner({ src, rule: RULE, fix: true });

		expect(result.errorCount).toBe(0);

		expect(getReplacements(result)).toMatchSnapshot();
	});
});