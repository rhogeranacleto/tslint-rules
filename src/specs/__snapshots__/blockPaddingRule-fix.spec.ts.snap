// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blockPadding Rule class A {
			constructor() {
				console.log(1);
			}
		} should throw error for line after method 1`] = `
"class A {
			constructor() {

				console.log(1);
			}
		}"
`;

exports[`blockPadding Rule class A {
			public b() {
				console.log(1);
			}
		} should throw error for line after method 1`] = `
"class A {
			public b() {

				console.log(1);
			}
		}"
`;

exports[`blockPadding Rule class A {
			public b() {

				console.log(1);

			}
		} should throw error for line after method 1`] = `
"class A {
			public b() {

				console.log(1);
			}
		}"
`;

exports[`blockPadding Rule done(() => {
		return 'maria';
	}) should throw error for line after method 1`] = `
"done(() => {

		return 'maria';
	})"
`;

exports[`blockPadding Rule done(() => {

			return 'maria';

		}) should throw error for line after method 1`] = `
"done(() => {

			return 'maria';
		})"
`;

exports[`blockPadding Rule for (let i = 0; i < 10; i++) {
		console.log(i);
	} should throw error for line after method 1`] = `
"for (let i = 0; i < 10; i++) {

		console.log(i);
	}"
`;

exports[`blockPadding Rule for (let i = 0; i < 10; i++) {

			console.log(i);

		} should throw error for line after method 1`] = `
"for (let i = 0; i < 10; i++) {

			console.log(i);
		}"
`;

exports[`blockPadding Rule for (let test in testes) {
		console.log(test);
	} should throw error for line after method 1`] = `
"for (let test in testes) {

		console.log(test);
	}"
`;

exports[`blockPadding Rule for (let test in testes) {

			console.log(test);

		} should throw error for line after method 1`] = `
"for (let test in testes) {

			console.log(test);
		}"
`;

exports[`blockPadding Rule for (let test of testes) {
		console.log(test);
	} should throw error for line after method 1`] = `
"for (let test of testes) {

		console.log(test);
	}"
`;

exports[`blockPadding Rule for (let test of testes) {

			console.log(test);

		} should throw error for line after method 1`] = `
"for (let test of testes) {

			console.log(test);
		}"
`;

exports[`blockPadding Rule function maria(){
		console.log('true');
	} should throw error for line after method 1`] = `
"function maria(){

		console.log('true');
	}"
`;

exports[`blockPadding Rule function maria(){

			console.log('true');

		} should throw error for line after method 1`] = `
"function maria(){

			console.log('true');
		}"
`;

exports[`blockPadding Rule if (true) {
			console.log(true);
		} should throw error for line after method 1`] = `
"if (true) {

			console.log(true);
		}"
`;

exports[`blockPadding Rule if (true) {
		console.log(true);
	} else {
		console.log(false)
	} should throw error for line after method 1`] = `
"if (true) {

		console.log(true);
	} else {

		console.log(false)
	}"
`;

exports[`blockPadding Rule if (true) {
		console.log(true);
	} else if (false) {
		console.log(false)
	} should throw error for line after method 1`] = `
"if (true) {

		console.log(true);
	} else if (false) {

		console.log(false)
	}"
`;

exports[`blockPadding Rule if (true) {

			console.log(true);
		} else {

			console.log(false)

		} should throw error for line after method 1`] = `
"if (true) {

			console.log(true);
		} else {

			console.log(false)
		}"
`;

exports[`blockPadding Rule if (true) {

			console.log(true);

			joao();


		} should throw error for line after method 1`] = `
"if (true) {

			console.log(true);

			joao();
		}"
`;

exports[`blockPadding Rule if (true) {

			console.log(true);

		} else {

			console.log(false)

		} should throw error for line after method 1`] = `
"if (true) {

			console.log(true);
		} else {

			console.log(false)
		}"
`;

exports[`blockPadding Rule if (true) {

			console.log(true);

		} else if (false) {

			console.log(false)

		} should throw error for line after method 1`] = `
"if (true) {

			console.log(true);
		} else if (false) {

			console.log(false)
		}"
`;

exports[`blockPadding Rule if (true) {

		console.log(true);
	} else {
		console.log(false)
	} should throw error for line after method 1`] = `
"if (true) {

		console.log(true);
	} else {

		console.log(false)
	}"
`;

exports[`blockPadding Rule it('jj', () => {
			expect(result.errorCount).toBe(1);
	}); should throw error for line after method 1`] = `
"it('jj', () => {

			expect(result.errorCount).toBe(1);
	});"
`;

exports[`blockPadding Rule it('jj', () => {

			expect(result.errorCount).toBe(1);

		}); should throw error for line after method 1`] = `
"it('jj', () => {

			expect(result.errorCount).toBe(1);
		});"
`;

exports[`blockPadding Rule try {
		console.log(0/0);
	} catch(e){
		console.log(0)
	} should throw error for line after method 1`] = `
"try {

		console.log(0/0);
	} catch(e){

		console.log(0)
	}"
`;

exports[`blockPadding Rule try {

			console.log(0/0);

		} catch(e){

			console.log(0)

		} should throw error for line after method 1`] = `
"try {

			console.log(0/0);
		} catch(e){

			console.log(0)
		}"
`;
